org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/resource/Storage2UriMapperImpl.java

```diff
@@ -216,9 +219,22 @@ protected boolean isHandled(IFolder folder) {
 
 	private Iterable<Pair<IStorage, IProject>> getStorages(/* @NonNull */ URI uri, IFile file) {
 		if (file == null || !file.isAccessible()) {
-			return contribution.getStorages(uri);
+			Iterable<Pair<IStorage, IProject>> result = contribution.getStorages(uri);
+			if (result == null || !result.iterator().hasNext()) {
+				 // Handle files external to the workspace. But check contribution first to be backwards compatible.
+				if (uri.isFile()) {
+					Path filePath = new Path(uri.toFileString());
+					IFileStore fileStore = EFS.getLocalFileSystem().getStore(filePath);
+					IFileInfo fileInfo = fileStore.fetchInfo();
+					if (fileInfo.exists() && !fileInfo.isDirectory()) {
+						return Collections.singletonList(
+								Tuples.<IStorage, IProject> create(new FileStoreStorage(fileStore, fileInfo, filePath), (IProject) null));
```

Comment:
Should we also override internalGetURI for FileStoreStorages?

Reply:
There is also https://github.com/eclipse/n4js/blob/master/plugins/org.eclipse.n4js.ts.ui/src/org/eclipse/n4js/ts/ui/navigation/URIBasedStorage.java

Would it be possible to use a similar impl? It appears to be slightly simpler, e.g. no EFS and stuff

Reply:
Ping

Reply:
I adapted "internalGetUri"